org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/outline/ModelBuilder.scala

```diff
@@ -0,0 +1,435 @@
+package org.scalaide.ui.internal.editor.outline
+/**
+ * ModelBuilder is the "heart" of Outline View. It maps scala source into a tree of nodes (model). Each node
+ * represent a scala entity, such as Class, Trait, Method and so on.
+ * This tree is rendered by TreeView later. Parsed tree is used to build the model. Which means not all
+ * type information is available.
+ * The model is built on the reconciler thread. Every time new model is generated it overwrites the previous
+ * one. There are two reasons why the model is mutable.
+ *
+ * 1. Despite of good model builder performance, applying model to TreeView could be expansive. For large file
+ * tens of thousands nodes are generated. Each node is translated to UI widget and all widget manipulation
+ * is done on UI thread, which may slow down the application.
+ *
+ * 2. The SWT TreeView has a state- user may expand/close some nodes. If we replace model, that state is lost.
+ *
+ * Using mutable model allows to update only nodes are modified.
+ */
+import scala.tools.nsc.Settings
+import scala.tools.nsc.interactive.CompilerControl
+import scala.tools.nsc.interactive.Global
+import scala.tools.nsc.reporters.ConsoleReporter
+import scala.reflect.internal.util.BatchSourceFile
+import scala.reflect.internal.util.SourceFile
+import scala.tools.nsc.interactive.Response
+import org.scalaide.core.compiler.IScalaPresentationCompiler
+import org.scalaide.logging.HasLogger
+import scala.collection.mutable.MutableList
+
+trait HasReturnType {
+  private var rt: Option[String] = None
+  def returnType = rt
+  def returnType_=(rt: Option[String]) = {
+    this.rt = rt
+  }
+}
+trait HasModifiers {
+  import scala.reflect.internal.Flags._
+  private[outline] var flags = 0L
+  def setFlags(f: Long) = {
+    flags = f
+  }
+  def isPrivate = (flags & PRIVATE) != 0
+  def isProtected = (flags & PROTECTED) != 0
+  def isTrait = (flags & TRAIT) != 0
+  def isParam = (flags & PARAM) != 0
+  def isSynthetic = (flags & SYNTHETIC) != 0
+  def isImplicit = (flags & IMPLICIT) != 0
+}
+sealed abstract class NodeKey
+
+sealed abstract class Node(val isLeaf: Boolean, val parent: ContainerNode) extends HasModifiers {
+  def name: String
+  var start: Int = 0
+  var end: Int = 0
+  def displayName = name
```

Comment:
I do not really like the `displayName` method. Some of the string conversion is done in here, the rest is done in `ModelBuild.buildTree`. I think it is more maintainable when everything is in one place. What about replacing `displayName` by a big pattern match?